File: week1_baseline/python/src/boukensha/logger.py
from __future__ import annotations

import json
import os
import re
import secrets
from datetime import datetime, timezone
from typing import Callable


class Logger:
    DEFAULT_SESSION_DIR = "sessions"

    def __init__(
        self,
        *,
        session_id: str | None = None,
        dir: str | None = None,
        log: str | None = None,
        snapshot: dict | None = None,
    ) -> None:
        self.session_id = session_id or self._generate_session_id()
        self.path = log or os.path.join(dir or self._default_dir(), f"{self.session_id}.jsonl")

        os.makedirs(os.path.dirname(self.path), exist_ok=True)
        self._log_io = open(self.path, "a", encoding="utf-8")
        self._subscribers: list[Callable[[dict], None]] = []
        self._write_log({"phase": "session_start", **(snapshot or {})})

    def turn(self, *, n: int) -> None:
        self._write_log({"phase": "turn", "n": n})

    def iteration(self, *, n: int, max: int) -> None:
        self._write_log({"phase": "iteration", "n": n, "max": max})

    def limit_reached(self, *, kind: str, n: int, max: int) -> None:
        self._write_log({"phase": "limit_reached", "kind": kind, "n": n, "max": max})

    def turn_end(self, *, reason: str, iterations: int, tokens=None) -> None:
        self._write_log({"phase": "turn_end", "reason": reason, "iterations": iterations, "tokens": tokens})

    def prompt(self, *, messages: list, tools: dict) -> None:
        self._write_log({
            "phase": "prompt",
            "message_count": len(messages),
            "messages": [self._serialize_message(m) for m in messages],
            "tool_count": len(tools),
            "tools": list(tools.keys()),
        })

    def tool_call(self, *, name: str, args: dict) -> None:
        self._write_log({"phase": "tool_call", "name": name, "args": args})

    def tool_result(self, *, name: str, result, ok: bool = True, error: str | None = None) -> None:
        self._write_log({"phase": "tool_result", "name": name, "result": str(result), "ok": ok, "error": error})

    def response(self, *, text, usage=None, stop_reason=None, task=None, backend=None) -> None:
        event = {
            "phase": "response",
            "text": str(text).strip(),
            "usage": usage,
            "stop_reason": stop_reason,
        }
        event.update(self._execution_metadata(task=task, backend=backend, usage=usage))
        self._write_log(event)

    def raw(self, *, data) -> None:
        import boukensha

        if not boukensha.is_debug():
            return

        self._write_log({"phase": "raw", "data": data})

    def subscribe(self, block: Callable[[dict], None]) -> None:
        self._subscribers.append(block)

    def close(self) -> None:
        self._log_io.close()

    def _default_dir(self) -> str:
        import boukensha

        return os.path.join(boukensha.config().dir, self.DEFAULT_SESSION_DIR)

    def _write_log(self, event: dict) -> None:
        payload = {
            **event,
            "session_id": self.session_id,
            "at": datetime.now().astimezone().isoformat(timespec="seconds"),
        }
        self._log_io.write(json.dumps(payload) + "\n")
        self._log_io.flush()
        for subscriber in self._subscribers:
            subscriber(event)

    def _generate_session_id(self) -> str:
        timestamp = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
        return f"{timestamp}-{secrets.token_hex(4)}"

    def _serialize_message(self, msg) -> dict:
        return {"role": msg.role, "content": msg.content}

    def _execution_metadata(self, *, task, backend, usage) -> dict:
        if task is None and backend is None and usage is None:
            return {}

        tokens = self._usage_tokens(usage)
        metadata = {
            "task": self._task_name(task),
            "provider": self._provider_name(backend),
            "model": getattr(backend, "model", None),
            "usage_unit": getattr(backend, "usage_unit", None),
            "usage_level": getattr(backend, "usage_level", None),
            "input_tokens": tokens["input"],
            "output_tokens": tokens["output"],
            "cost_usd": self._estimate_cost(backend, tokens),
        }
        return {k: v for k, v in metadata.items() if v is not None}

    def _task_name(self, task):
        if task is None:
            return None
        if hasattr(task, "task_name"):
            return task.task_name()
        return str(task)

    def _provider_name(self, backend):
        if backend is None:
            return None
        name = type(backend).__name__
        return re.sub(r"([a-z0-9])([A-Z])", r"\1_\2", name).lower()

    def _usage_tokens(self, usage) -> dict:
        usage = usage or {}
        return {
            "input": self._first_integer(
                usage, "input_tokens", "prompt_tokens", "promptTokenCount", "prompt_eval_count"
            ),
            "output": self._first_integer(
                usage, "output_tokens", "completion_tokens", "candidatesTokenCount", "eval_count"
            ),
        }

    def _first_integer(self, data: dict, *keys: str):
        try:
            for key in keys:
                value = data.get(key)
                if value is not None:
                    return int(value)
            return None
        except (TypeError, ValueError):
            return None

    def _estimate_cost(self, backend, tokens):
        if backend is None or not hasattr(backend, "estimate_cost"):
            return None
        if tokens["input"] is None or tokens["output"] is None:
            return None
        return backend.estimate_cost(input_tokens=tokens["input"], output_tokens=tokens["output"])

File: week2_capable/dotnet/src/Boukensha.Core/Logger.cs
using System.Security.Cryptography;
using System.Text.Json;

namespace Boukensha.Core;

public sealed class Logger : IDisposable
{
    private readonly StreamWriter _writer;
    private readonly string _sessionId;
    private readonly List<Action<IReadOnlyDictionary<string, object?>>> _subscribers = [];
    private readonly Lock _lock = new();

    public string Path { get; }

    public Logger(string dir, string? sessionId = null, string? log = null, IReadOnlyDictionary<string, object?>? snapshot = null)
    {
        _sessionId = sessionId ?? GenerateSessionId();
        Path = log ?? System.IO.Path.Combine(dir, $"{_sessionId}.jsonl");
        Directory.CreateDirectory(System.IO.Path.GetDirectoryName(Path)!);
        _writer = new StreamWriter(new FileStream(Path, FileMode.Append, FileAccess.Write, FileShare.ReadWrite)) { AutoFlush = true };

        var start = new Dictionary<string, object?> { ["phase"] = "session_start" };
        if (snapshot is not null)
        {
            foreach (var (key, value) in snapshot) start[key] = value;
        }
        WriteLog(start);
    }

    public void Turn(int n) => WriteLog(new() { ["phase"] = "turn", ["n"] = n });

    public void Iteration(int n, int max) => WriteLog(new() { ["phase"] = "iteration", ["n"] = n, ["max"] = max });

    public void LimitReached(string kind, int n, int max) =>
        WriteLog(new() { ["phase"] = "limit_reached", ["kind"] = kind, ["n"] = n, ["max"] = max });

    public void TurnEnd(string reason, int iterations, int? tokens = null) =>
        WriteLog(new() { ["phase"] = "turn_end", ["reason"] = reason, ["iterations"] = iterations, ["tokens"] = tokens });

    public void Prompt(IReadOnlyList<Message> messages, IReadOnlyDictionary<string, ToolDefinition> tools, int contextWindow) =>
        WriteLog(new()
        {
            ["phase"] = "prompt",
            ["messages"] = messages.Select(SerializeMessage).ToList(),
            ["message_count"] = messages.Count,
            ["tool_count"] = tools.Count,
            ["tools"] = tools.Keys.ToList(),
            ["context_window"] = contextWindow,
        });

    public void ToolCatalog(IReadOnlyDictionary<string, ToolDefinition> tools) =>
        WriteLog(new()
        {
            ["phase"] = "tool_catalog",
            ["tools"] = tools.Values.Select(t => new Dictionary<string, object?>
            {
                ["name"] = t.Name,
                ["description"] = t.Description,
                ["parameters"] = t.Parameters.ToDictionary(p => p.Key, p => new Dictionary<string, object?>
                {
                    ["type"] = p.Value.Type,
                    ["description"] = p.Value.Description,
                }),
            }).ToList(),
        });

    public void Compaction(int before, int dropped, int contextWindow) =>
        WriteLog(new() { ["phase"] = "compaction", ["before"] = before, ["dropped"] = dropped, ["context_window"] = contextWindow });

    public void ToolCall(string name, IReadOnlyDictionary<string, object?> args, string task) =>
        WriteLog(new() { ["phase"] = "tool_call", ["name"] = name, ["args"] = args, ["task"] = task });

    public void ToolResult(string name, string result, string task, int durationMs, bool ok = true, string? error = null) =>
        WriteLog(new()
        {
            ["phase"] = "tool_result",
            ["name"] = name,
            ["result"] = result,
            ["task"] = task,
            ["duration_ms"] = durationMs,
            ["ok"] = ok,
            ["error"] = error,
        });

    public void Response(string text, IReadOnlyDictionary<string, object?>? usage, string? stopReason, string? task, string? backend, double? costUsd, int durationMs) =>
        WriteLog(new()
        {
            ["phase"] = "response",
            ["text"] = text,
            ["usage"] = usage,
            ["stop_reason"] = stopReason,
            ["task"] = task,
            ["provider"] = backend,
            ["cost_usd"] = costUsd,
            ["duration_ms"] = durationMs,
        });

    public void Reasoning(string text, bool redacted = false) =>
        WriteLog(new() { ["phase"] = "reasoning", ["text"] = text, ["redacted"] = redacted });

    public void Plan(string text) => WriteLog(new() { ["phase"] = "plan", ["text"] = text });

    public void Subscribe(Action<IReadOnlyDictionary<string, object?>> handler)
    {
        lock (_lock) _subscribers.Add(handler);
    }

    public void Dispose() => _writer.Dispose();

    private static Dictionary<string, object?> SerializeMessage(Message message) => new()
    {
        ["role"] = message.Role,
        ["content"] = message.Content.IsText ? message.Content.Text : message.Content.Blocks,
    };

    private void WriteLog(Dictionary<string, object?> evt)
    {
        evt["session_id"] = _sessionId;
        evt["at"] = DateTimeOffset.Now.ToString("yyyy-MM-ddTHH:mm:sszzz");
        List<Action<IReadOnlyDictionary<string, object?>>> subscribersSnapshot;
        lock (_lock)
        {
            _writer.WriteLine(JsonSerializer.Serialize(evt));
            subscribersSnapshot = [.. _subscribers];
        }
        foreach (var subscriber in subscribersSnapshot) subscriber(evt);
    }

    public static string GenerateSessionId()
    {
        var timestamp = DateTime.UtcNow.ToString("yyyyMMddTHHmmssZ");
        var suffix = Convert.ToHexString(RandomNumberGenerator.GetBytes(4)).ToLowerInvariant();
        return $"{timestamp}-{suffix}";
    }
}
